### PR TITLE
Wasm upgrade from main

### PR DESCRIPTION
**Pull request by nghuyenthevinh2000**
## Summary of changes

Clean updating wasm from main branch.

cc: @fragwuerdig

This is work from: #49 

Change:
1. added in benchmarking tool to calculate gas
2. update Dockerfile
3. refactor test

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Updated API documentation (client/lcd/swagger-ui/swagger.yaml)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

**Comment by nghuyenthevinh2000** [comment]
I find the gas used by wasmvm after upgrade to v1.0.0 is pretty staggering. I put a quick comparison between v0.16.6 and v1.0.0 gas consumption here:
1. v0.16.6: 141633
2. v1.0.0: 1874816400

that is x10000

@fragwuerdig @ZaradarBH

**Comment by fragwuerdig** [comment]
@nghuyenthevinh2000 did you adjust the gas multiplier in this branch? Please look in my original #49 and try to cherry pick as much as you can from there. Old multiplier was 100. New one is 140,000,000.

**Comment by nghuyenthevinh2000** [comment]
> @nghuyenthevinh2000 did you adjust the default gas multiplier in this branch?
> 
> Please look in my original #49 and try to cherry pick as much as you can from there.
> 
> Old gas multiplier was 100. New one is about 140,000,000.

I find this very interesting actually. Like, why wasmvm just increases its gas price so huge like that.

**Comment by nghuyenthevinh2000** [comment]
@fragwuerdig 

I really want to bring our x/wasm closer to CosmWasm x/wasm so that one day we can just use x/wasm and save us from the pain of maintaining this module. After comparing the two, I realize that there are a lot of functions with same functionality but different name.

**Comment by fragwuerdig** [comment]
@nghuyenthevinh2000 so you are suggesting to refactor towards wasmd implementation. We can definitely work on that. Could we have a separate issue for that an just focus on getting the upgrade done for now? Or do you think refactoring should be done asap?

**Comment by nghuyenthevinh2000** [comment]
> @nghuyenthevinh2000 so you are suggesting to refactor towards wasmd implementation? We can definitely work on that. Could we have a separate issue for that an just focus on getting the upgrade done for now? Or do you think refactoring should be done asap?

Overtime I would say, right now, I am trying to get the testing code of wasmd into ours.

**Comment by ZaradarBH** [comment]
Need to fix the broken tests :)

File: x/wasm/keeper/api.go [hunk 1]
```diff
@@ -8,12 +8,13 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/terra-money/core/x/wasm/types"
+
```
Comment (ZaradarBH) [inline x/wasm/keeper/api.go:11]: Go fumpt :P

**Comment by nghuyenthevinh2000** [comment]
I have been playing around with gas to try to see if it can be calculated. The wasmVM gas calculation behaves differently from last version. After understanding the flow of gas calculation in wasmVM, my conclusion is that however wasmVM behaves we should only concern about GasMultiplier.

Name of the game: guess GasMultiplier so that it breaks as less test as possible.
Some tests are guaranteed to break: 
* TestGasCostOnQuery
* TestLimitRecursiveQueryGas

Because they check for exact gas which will always break everytime GasMultiplier is changed.

I find that GasMultiplier = 15_000_000 breaks the least amount of gas test

**Review by ZaradarBH — APPROVED** [review]

**Comment by nghuyenthevinh2000** [comment]
I will write a separate pr for wasm upgrade testing

**Pull request merged**